### PR TITLE
🎨(backend) use REST style api path for RoomUser and RoomGroup

### PR DESCRIPTION
**Pull request by kernicPanel**
## Purpose

We want to use api urls in the form `/api/rooms/{room.id}/users/{user.id}/`

## Proposal

- [x] Replace RoomUserViewSet by actions in RoomViewSet.
- [x] Replace RoomGroupViewSet by actions in RoomViewSet.

### CODE REVIEW

File: src/magnify/apps/core/api.py [hunk 1]
```diff
@@ -234,3 +202,125 @@ def perform_create(self, serializer):
...
+        methods=["get", "put", "delete"],
+        detail=True,
+        url_path="users/(?P<user_pk>[^/.]+)",
+        permission_classes=[IsRoomAdministrator],
+    )
+    # pylint: disable=unused-argument,invalid-name
+    def user(self, request, user_pk, pk=None):
+        """Get, update, or delete a user of a room."""
+        room = self.get_object()
+        room_user = RoomUser.objects.get(room=room, user__pk=user_pk)
```
Comment (sampaccoud) [inline src/magnify/apps/core/api.py:240]: now I know why it should be called `RoomUserRelation`: it took me 1 minute to understand that this was not a user... The reader a RoomUser is a user in a room...

File: src/magnify/apps/core/api.py [hunk 2]
```diff
@@ -234,3 +202,125 @@ def perform_create(self, serializer):
...
+
+        if request.method == "PUT":
+            user = User.objects.get(id=user_pk)
+            serializer = RoomUserSerializer(
+                room_user,
+                data={
+                    "room": room.id,
+                    "user": user.id,
+                    "is_administrator": request.data.get("is_administrator"),
+                },
```
Comment (sampaccoud) [inline src/magnify/apps/core/api.py:258]: In fact it seems you are getting the user by its pk only to get its id, and the room only to get its id. 
```suggestion
            serializer = RoomUserSerializer(
                room_user,
                data={
                    "room": pk,
                    "user": user_pk,
                    "is_administrator": request.data.get("is_administrator"),
                },
```

File: src/magnify/apps/core/api.py [hunk 3]
```diff
@@ -234,3 +202,125 @@ def perform_create(self, serializer):
...
+    @action(
+        methods=["get", "put", "delete"],
+        detail=True,
+        url_path="users/(?P<user_pk>[^/.]+)",
+        permission_classes=[IsRoomAdministrator],
+    )
+    # pylint: disable=unused-argument,invalid-name
+    def user(self, request, user_pk, pk=None):
+        """Get, update, or delete a user of a room."""
+        room = self.get_object()
```
Comment (sampaccoud) [inline src/magnify/apps/core/api.py:239]: I think you don't need this line (you can directly use pk for the PUT below)

File: src/magnify/apps/core/api.py [hunk 4]
```diff
@@ -234,3 +202,125 @@ def perform_create(self, serializer):
...
+    # pylint: disable=unused-argument,invalid-name
+    def users(self, request, pk=None):
+        """Adds a user in a room."""
+        room = self.get_object()
+        user = User.objects.get(id=request.data.get("user"))
+        serializer = RoomUserSerializer(
+            data={
+                "room": room.id,
+                "user": user.id,
+                "is_administrator": request.data.get("is_administrator"),
```
Comment (sampaccoud) [inline src/magnify/apps/core/api.py:221]:
```suggestion
        serializer = RoomUserSerializer(
            data={
                "room": pk,
                "user": request.data.get("user"),
                "is_administrator": request.data.get("is_administrator"),
```

**Review by DarkPitoune — APPROVED** [review]

File: src/magnify/apps/core/api.py (hunk 1 above)
Comment (kernicPanel) [reply src/magnify/apps/core/api.py:240]: Relation still confuses me… Maybe RoomUserRole ?

File: src/magnify/apps/core/api.py (hunk 2 above)
Comment (kernicPanel) [reply src/magnify/apps/core/api.py:258]: For the user, I agree!
Four the room, too, but apparently we have to call `self.get_object()` for django to check for `has_object_permission`. And calling it directly will require us to get it anyway.

File: src/magnify/apps/core/api.py (hunk 3 above)
Comment (kernicPanel) [reply src/magnify/apps/core/api.py:239]: Same here.

File: src/magnify/apps/core/api.py (hunk 4 above)
Comment (kernicPanel) [reply src/magnify/apps/core/api.py:221]: And here.

File: src/magnify/apps/core/api.py (hunk 1 above)
Comment (sampaccoud) [reply src/magnify/apps/core/api.py:240]: or `RoomUserAccess` like in [marsha](https://github.com/openfun/marsha)?

File: src/magnify/apps/core/api.py (hunk 2 above)
Comment (sampaccoud) [reply src/magnify/apps/core/api.py:258]: Ok good point!

File: src/magnify/apps/core/api.py (hunk 1 above)
Comment (kernicPanel) [reply src/magnify/apps/core/api.py:240]: Yes, good idea.
Let's do this in a dedicated PR.

**Pull request merged**